File: SDK Examples/Examples/VB.Net/WaveformGenerator.vb
Imports System.Collections
Imports System.IO
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Drawing.Text
Imports System.Text
Imports System.Windows.Forms
Imports System.Drawing.Drawing2D
Imports Advantech.BioDaq

Enum WaveformStyle
   sine
   square
   triangle
End Enum

Class WaveformGenerator
#Region "fields"
   Public m_pointPerPeriod As UInteger
#End Region

#Region "ctor and dtor"
   Public Sub New(ByVal pointPerPeriod As UInteger)
      m_pointPerPeriod = pointPerPeriod
   End Sub
#End Region

#Region "methods"

   Public Function GetOnePoint(ByVal style As WaveformStyle, ByVal index As Integer, ByVal highLevel As [Double], ByVal lowLevel As [Double]) As Double
      Dim amplitude As Double = (highLevel - lowLevel) / 2
      Dim offset As Double = (highLevel + lowLevel) / 2
      Dim data As Double = 0
      System.Diagnostics.Debug.Assert(index >= 0 AndAlso index <= m_pointPerPeriod)
      Select Case style
         Case WaveformStyle.sine
            data = amplitude * Math.Sin(index * 2.0 * Math.PI / m_pointPerPeriod) + offset
            Exit Select
         Case WaveformStyle.triangle
            If index < (m_pointPerPeriod / 4.0) Then
               data = amplitude * (index / (m_pointPerPeriod / 4.0)) + offset
            ElseIf index < 3 * (m_pointPerPeriod / 4.0) Then
               data = amplitude * ((2.0 * (m_pointPerPeriod / 4.0) - index) / (m_pointPerPeriod / 4.0)) + offset
            Else
               data = amplitude * ((index - m_pointPerPeriod) / (m_pointPerPeriod / 4.0)) + offset
            End If
            Exit Select
         Case WaveformStyle.square
            If (index >= 0) AndAlso (index < (m_pointPerPeriod \ 2)) Then
               data = amplitude * 1 + offset
            Else
               data = amplitude * (-1) + offset
            End If
            Exit Select
         Case Else
            Exit Select
      End Select
      Return data
   End Function

   Public Sub GetOnePeriod(ByVal waveformBuffer As [Double](), ByVal style As WaveformStyle, ByVal highLevel As [Double], ByVal lowLevel As [Double])
      For i As Integer = 0 To m_pointPerPeriod - 1
         waveformBuffer(i) = GetOnePoint(style, i, highLevel, lowLevel)
      Next
   End Sub

   Public Sub GetMultipleWaveform(ByVal waveform As [Double](), ByVal waveCount As Integer, ByVal styles As WaveformStyle(), ByVal highLevel As [Double](), ByVal lowLevel As [Double]())
      For i As Integer = 0 To m_pointPerPeriod - 1
         For j As Integer = 0 To waveCount - 1
            waveform(j + i * waveCount) = GetOnePoint(styles(j), i, highLevel(j), lowLevel(j))
         Next
      Next
   End Sub
#End Region
End Class

File: SDK Examples/Examples/C#_Console/WaveformGenerator.cs
using System;
using System.Collections;
using System.IO;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Drawing.Text;
using System.Text;
using System.Windows.Forms;
using System.Drawing.Drawing2D;
using Automation.BDaq;



    enum WaveformStyle { sine, square,triangle};
    class WaveformGenerator
    {
        #region fields
       public uint m_pointPerPeriod;
        #endregion

        #region ctor and dtor
        public WaveformGenerator(uint pointPerPeriod)
        {
            m_pointPerPeriod = pointPerPeriod;
        }
        #endregion

        #region methods

        public double GetOnePoint(WaveformStyle style, int index, Double highLevel, Double lowLevel)
        {
            double amplitude = (highLevel - lowLevel) / 2;
            double offset = (highLevel + lowLevel) / 2;
            double data = 0;
            System.Diagnostics.Debug.Assert(index >= 0 && index <= m_pointPerPeriod);
            switch (style)
            {
            case WaveformStyle.sine:
                data = amplitude * Math.Sin(index * 2.0 * Math.PI / m_pointPerPeriod) + offset;
                break;
            case WaveformStyle.triangle:
                if (index < (m_pointPerPeriod / 4.0))
                {
                    data = amplitude * (index / (m_pointPerPeriod / 4.0)) + offset;
                }
                else if (index < 3 * (m_pointPerPeriod / 4.0))
                {
                    data = amplitude * ((2.0 * (m_pointPerPeriod / 4.0) - index) / (m_pointPerPeriod / 4.0)) + offset;
                }
                else
                {
                    data = amplitude * ((index - m_pointPerPeriod) / (m_pointPerPeriod / 4.0)) + offset;
                }
                break;
            case WaveformStyle.square:
                if ((index >= 0) && (index < (m_pointPerPeriod / 2)))
                {
                    data = amplitude * 1 + offset;
                }
                else
                {
                    data = amplitude * (-1) + offset;
                }
                break;
             default:
                 break;
            }
            return data;
        }

        public void GetOnePeriod(Double[] waveformBuffer,WaveformStyle style, Double highLevel, Double lowLevel)
        {
            for (int i = 0; i < m_pointPerPeriod; i++ )
            {
                waveformBuffer[i] = GetOnePoint(style, i, highLevel, lowLevel);
            }
        }

        public void GetMultipleWaveform(Double[] waveform,int waveCount, WaveformStyle[] styles,Double[] highLevel, Double[] lowLevel)
        {
            for (int i = 0; i < m_pointPerPeriod;i++ )
            {
                for (int j = 0; j < waveCount; j++)
                {
                    waveform[j + i * waveCount] = GetOnePoint(styles[j], i, highLevel[j], lowLevel[j]);
                }
            }
        }
        #endregion
    }

